serverless/internal/storage/webstorage/fs.go

```diff
@@ -0,0 +1,290 @@
+// Copyright 2021 Google LLC. All Rights Reserved.
+//
+// Licensed under the Apache License, Version 2.0 (the "License");
+// you may not use this file except in compliance with the License.
+// You may obtain a copy of the License at
+//
+//     http://www.apache.org/licenses/LICENSE-2.0
+//
+// Unless required by applicable law or agreed to in writing, software
+// distributed under the License is distributed on an "AS IS" BASIS,
+// WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+// See the License for the specific language governing permissions and
+// limitations under the License.
+
+// +build wasm
+
+// Package webstorage provides a simple log storage implementation based on webstorage.
+// It only really makes sense for wasm targets.
+package webstorage
+
+import (
+	"crypto/sha256"
+	"errors"
+	"fmt"
+	"os"
+	"path/filepath"
+	"strconv"
+	"strings"
+	"syscall/js"
+
+	"github.com/golang/glog"
+	"github.com/google/trillian-examples/formats/log"
+	"github.com/google/trillian-examples/serverless/api"
+	"github.com/google/trillian-examples/serverless/api/layout"
+	"github.com/google/trillian-examples/serverless/internal/storage"
+)
+
+// Storage is a serverless storage implementation which uses webstorage entries to store tree state.
+// The storage key format is:
+//  <root>/leaves/aa/bb/cc/ddeeff...
+//  <root>/leaves/pending/aabbccddeeff...
+//  <root>/seq/aa/bb/cc/ddeeff...
+//  <root>/tile/<level>/aa/bb/ccddee...
+//  <root>/checkpoint
+//
+// The functions on this struct are not thread-safe.
+type Storage struct {
+	// root is the prefix of all webstorage keys which store tree data.
+	root string
+	// nextSeq is a hint to the Sequence func as to what the next available
+	// sequence number is to help performance.
+	// Note that nextSeq may be <= than the actual next available number, but
+	// never greater.
+	nextSeq uint64
+	// checkpoint is the latest known checkpoint of the log.
+	checkpoint log.Checkpoint
+}
+
+const leavesPendingPathFmt = "leaves/pending/%0x"
+
+func getStorage() js.Value {
+	return js.Global().Get("sessionStorage")
+}
+
+func exists(k string) bool {
+	_, err := get(k)
+	return err == nil
+}
+
+func get(k string) ([]byte, error) {
+	v := getStorage().Call("getItem", js.ValueOf(k))
+	if js.Null().Equal(v) {
+		return nil, os.ErrNotExist
+	}
+	return []byte(v.String()), nil
+}
+
+func set(k string, v []byte) error {
+	getStorage().Call("setItem", js.ValueOf(k), js.ValueOf(string(v)))
+	// TODO(al) read back and check
+	return nil
+}
+
+func createExclusive(k string, v []byte) error {
+	if exists(k) {
+		return os.ErrExist
+	}
+	getStorage().Call("setItem", js.ValueOf(k), js.ValueOf(string(v)))
+	// TODO(al) read back and check
+	return nil
+}
+
+// Load returns a Storage instance initialised from webstorage prefixed at root.
+func Load(root string, checkpoint *log.Checkpoint) (*Storage, error) {
```

Comment:
Passing the checkpoint into this method feels odd to me. It should be stored in the same root filesystem. Asking the client of this method to do the dance to call `cp, _ := ReadCheckpoint(foo); s, _ := Load(foo, cp)` seems risky and unclear as an API.

Reply:
Perhaps a method like `Exists(root)` or `Verify(root)` would be better. The client flow would then be:

```
var s Storage
if !webstorage.Exists(root) {
  s = webstorage.Create(root)
} else {
  s = webstorage.Load(root)
}
```

The implementation of Exists could just check there is a checkpoint there.

Reply:
This is the same as the `storage/fs` package it was forked from.

The reason for this is because the storage should not have to know about signatures/verifiers - all of that is handled at the caller's level (both opening and passing in here, as well as signing and persisting new CPs for larger trees).